File: C#/SyntacticalAnalyzer.cs
using System;

namespace Compilador
{
    class SyntacticallAnalyzer
    {
        private readonly LexicalAnalyzer lexicalAnalyzer;
        private Token token;
        private int line;

        public SyntacticallAnalyzer(LexicalAnalyzer lexicalAnalyzer)
        {
            this.lexicalAnalyzer = lexicalAnalyzer;
            token = lexicalAnalyzer.GetToken();
            line = 1;
        }

        private void advance()
        {
            token = lexicalAnalyzer.GetToken();
            line = lexicalAnalyzer.Line;
        }

        private bool eat(TokenType tokenType, bool throwsException = true)
        {
            if (token?.Type == tokenType)
            {
                advance();
                return true;
            }
            if (throwsException) throw new Exception($"Expected token '{tokenType}' not found. Found '{token?.Type}' instead on line {line}.");
            return false;
        }

        public void program()
        {
            eat(TokenType.PROGRAM);
            declList();
            stmtList();
            eat(TokenType.END);
        }

        private void declList()
        {
            while (eat(TokenType.INT, throwsException: false) || eat(TokenType.STRING, throwsException: false)) // Type
            {
                do
                {
                    eat(TokenType.IDENTIFIER);
                } while (eat(TokenType.COMMA, throwsException: false));
                eat(TokenType.SEMICOLON);
            }
        }

        private bool stmtList(bool throwsException = true)
        {
            do
            {
                if (eat(TokenType.IDENTIFIER, throwsException: false)) // Assign Stmt
                {
                    eat(TokenType.ASSIGN);
                    simpleExpr();
                    eat(TokenType.SEMICOLON);
                }
                else if (eat(TokenType.IF, throwsException: false)) // If stmt
                {
                    expression();
                    eat(TokenType.THEN);
                    stmtList();
                    if (eat(TokenType.ELSE, throwsException: false))
                    {
                        stmtList();
                    }
                    eat(TokenType.END);
                }
                else if (eat(TokenType.DO, throwsException: false)) // Do while
                {
                    stmtList();
                    eat(TokenType.WHILE);
                    expression();
                    eat(TokenType.END);
                }
                else if (eat(TokenType.SCAN, throwsException: false)) // scan();
                {
                    eat(TokenType.PARENTHESES_OPEN);
                    eat(TokenType.IDENTIFIER);
                    eat(TokenType.PARENTHESES_CLOSE);
                    eat(TokenType.SEMICOLON);
                }
                else if (eat(TokenType.PRINT, throwsException: false)) // print();
                {
                    eat(TokenType.PARENTHESES_OPEN);
                    if (eat(TokenType.LITERAL, throwsException: false)) { }
                    else simpleExpr();
                    eat(TokenType.PARENTHESES_CLOSE);
                    eat(TokenType.SEMICOLON);
                }
                else if (throwsException)
                {
                    throw new Exception($"Unexpected token '{token.Type}' found on line {line}.");
                }
                else
                {
                    return false;
                }
            } while (stmtList(throwsException: false));
            return false;
        }

        private void expression()
        {
            simpleExpr();
            if (eat(TokenType.EQUALS, throwsException: false)
                || eat(TokenType.GREATER, throwsException: false)
                || eat(TokenType.GREATERTHAN, throwsException: false)
                || eat(TokenType.LESS, throwsException: false)
                || eat(TokenType.LESSTHAN, throwsException: false)
                || eat(TokenType.DIFFERENT, throwsException: false))
            {
                simpleExpr();
            }
        }

        private void simpleExpr()
        {
            if (eat(TokenType.EXCLAMATION, throwsException: false) || eat(TokenType.SUBTRACTION, throwsException: false)) { }
            factor();
            mulop();
            addop();
        }

        private void mulop()
        {
            if (eat(TokenType.MULTIPLICATION, throwsException: false)
                || eat(TokenType.DIVISION, throwsException: false)
                || eat(TokenType.AND, throwsException: false))
            {
                factor();
                mulop();
            }
        }

        private void addop()
        {
            if (eat(TokenType.ADDITION, throwsException: false)
                || eat(TokenType.SUBTRACTION, throwsException: false)
                || eat(TokenType.OR, throwsException: false))
            {
                factor();
                mulop();
                addop();
            }
        }

        private void factor()
        {
            if (eat(TokenType.IDENTIFIER, throwsException: false)
                || eat(TokenType.INT_CONSTANT, throwsException: false)
                || eat(TokenType.LITERAL, throwsException: false))
            {
                return;
            }
            if (eat(TokenType.PARENTHESES_OPEN, throwsException: false))
            {
                expression();
                eat(TokenType.PARENTHESES_CLOSE);
                return;
            }
            throw new Exception($"Unexpected token found '{token.Type}' on line {line}. Expected {TokenType.IDENTIFIER}, {TokenType.INT_CONSTANT}, {TokenType.LITERAL} or {TokenType.PARENTHESES_OPEN}");
        }
    }
}

File: SyntacticalAnalyzer.java
public class SyntacticalAnalyzer {
    private LexicalAnalyzer lexicalAnalyzer;
    private Token token;
    private Token lastToken;
    private int line;
    public boolean Error = false;

    public SyntacticalAnalyzer(LexicalAnalyzer lexicalAnalyzer) throws Exception
    {
        this.lexicalAnalyzer = lexicalAnalyzer;
        this.token = lexicalAnalyzer.GetToken();
        this.line = 1;
        program();
    }

    private void advance() throws Exception
    {
        this.lastToken = this.token;
        this.token = lexicalAnalyzer.GetToken();
        this.line = lexicalAnalyzer.Line;
    }

    private boolean eatIf(TokenType tokenType, boolean checkType) throws Exception
    {
        if (this.token != null && this.token.Type == tokenType)
        {
            if(checkType && this.token.Type == TokenType.IDENTIFIER && this.token.SemanticType == SemanticTokenType.NULL){
                throw new Exception("Using undeclared variable '"+token.Value+"'on line " + line);
            }
            advance();
            return true;
        }
        return false;
    }

    private boolean eat(TokenType tokenType, boolean checkType) throws Exception
    {
        if (this.token != null && this.token.Type == tokenType)
        {
            if(checkType && this.token.Type == TokenType.IDENTIFIER && this.token.SemanticType == SemanticTokenType.NULL){
                throw new Exception("Using undeclared variable '"+token.Value+"'on line " + line);
            }
            advance();
            return true;
        }
        throw new Exception("Expected token '"+ tokenType +"' not found. Found '"+token.Type+"' instead on line "+line+".");
    }

    public void program() throws Exception
    {
        eat(TokenType.PROGRAM, false);
        declList();
        stmtList(true);
        eat(TokenType.END, false);
        eat(TokenType.EOF, false);
    }

    private void declList() throws Exception
    {
        while (eatIf(TokenType.INT, false) || eatIf(TokenType.STRING, false)) // Type
        {
            SemanticTokenType type = lastToken.getSemanticTypeFromType();
            do
            {
                eat(TokenType.IDENTIFIER, false);
                if(lexicalAnalyzer.SymbolTable.get(lastToken.Value).SemanticType != SemanticTokenType.NULL) {
                    throw new Exception("Semantic Exception: " + lastToken.Value + " on line "+ line +" is already declared");
                }
                lastToken.SemanticType = type;
            } while (eatIf(TokenType.COMMA, false));
            eat(TokenType.SEMICOLON, false);
        }
    }

    private boolean stmtList(boolean throwsException) throws Exception
    {
        do
        {
            if (eatIf(TokenType.IDENTIFIER, true)) // Assign Stmt
            {
                SemanticTokenType tokenType = lastToken.SemanticType;
                eat(TokenType.ASSIGN, false);
                SemanticTokenType exprType = simpleExpr();
                if(tokenType != exprType) {
                    throw new Exception("Semantic Error: Expression ("+tokenType+") and Variable ("+exprType+") have incompatible type on line " + line);
                }
                eat(TokenType.SEMICOLON, false);
            }
            else if (eatIf(TokenType.IF, false)) // If stmt
            {
                expression();
                eat(TokenType.THEN, false);
                stmtList(true);
                if (eatIf(TokenType.ELSE, false))
                {
                    stmtList(true);
                }
                eat(TokenType.END, false);
            }
            else if (eatIf(TokenType.DO, false)) // Do while
            {
                stmtList(true);
                eat(TokenType.WHILE, false);
                expression();
                eat(TokenType.END, false);
            }
            else if (eatIf(TokenType.SCAN, false)) // scan();
            {
                eat(TokenType.PARENTHESES_OPEN, false);
                eat(TokenType.IDENTIFIER, true);
                eat(TokenType.PARENTHESES_CLOSE, false);
                eat(TokenType.SEMICOLON, false);
            }
            else if (eatIf(TokenType.PRINT, false)) // print();
            {
                eat(TokenType.PARENTHESES_OPEN, false);
                if (eatIf(TokenType.LITERAL, false)) { }
                else simpleExpr();
                eat(TokenType.PARENTHESES_CLOSE, false);
                eat(TokenType.SEMICOLON, false);
            }
            else if (throwsException) {
                throw new Exception("Unexpected token '"+token.Type+"' found on line "+line+".");
            }
            else {
                return false;
            }
        } while (stmtList(false));
        return false;
    }

    private SemanticTokenType expression() throws Exception
    {
        SemanticTokenType type = simpleExpr();
        if (eatIf(TokenType.EQUALS, false)
            || eatIf(TokenType.GREATER, false)
            || eatIf(TokenType.GREATERTHAN, false)
            || eatIf(TokenType.LESS, false)
            || eatIf(TokenType.LESSTHAN, false)
            || eatIf(TokenType.DIFFERENT, false))
        {
            simpleExpr();
        }
        return type;
    }

    private SemanticTokenType simpleExpr() throws Exception
    {
        SemanticTokenType type = factor();
        mulop();
        addop();
        return type;
    }

    private void mulop() throws Exception
    {
        if (eatIf(TokenType.MULTIPLICATION, false)
            || eatIf(TokenType.DIVISION, false)
            || eatIf(TokenType.AND, false))
        {
            factor();
            mulop();
        }
    }

    private void addop() throws Exception
    {
        if (eatIf(TokenType.ADDITION, false)
            || eatIf(TokenType.SUBTRACTION, false)
            || eatIf(TokenType.OR, false))
        {
            factor();
            mulop();
            addop();
        }
    }

    private SemanticTokenType factor() throws Exception
    {
        if (eatIf(TokenType.NOT, false) || eatIf(TokenType.SUBTRACTION, false)) { }
        if (eatIf(TokenType.IDENTIFIER, true)
            || eatIf(TokenType.INT_CONSTANT, false)
            || eatIf(TokenType.LITERAL, false))
        {
            if(lastToken.Type == TokenType.IDENTIFIER) return lastToken.SemanticType;
            else return lastToken.getSemanticTypeFromType();
        }
        if (eatIf(TokenType.PARENTHESES_OPEN, false))
        {
            SemanticTokenType type = expression();
            eat(TokenType.PARENTHESES_CLOSE, false);
            return type;
        }
        throw new Exception("Unexpected token found '"+token.Type+"' on line "+line+". Expected "+TokenType.IDENTIFIER+", "+TokenType.INT_CONSTANT+", "+TokenType.LITERAL+" or "+TokenType.PARENTHESES_OPEN);
    }
}